File: Walmart/seed.py
"""
Seeds the Walmart products collection from the public DummyJSON API.

Usage:
    python seed.py                    # add new products, skip duplicates
    python seed.py --clear            # delete all products first
    python seed.py --count=50         # how many to fetch (max 100)
    python seed.py --skip=50          # offset (default 50 so we don't overlap with Amazon's seed)

Reads MONGO_CONN from .env.
Requires: requests, mongoengine, python-dotenv.
"""

import os
import sys
from datetime import datetime
import requests
from dotenv import load_dotenv
from mongoengine import connect, NotUniqueError

from Models.product_model import Product

load_dotenv()


def parse_arg(args, name, default):
    for a in args:
        if a.startswith(f"--{name}="):
            return a.split("=", 1)[1]
    return default


def transform(p):
    brand = p.get("brand") or "Generic"
    return Product(
        name=p["title"],
        description=p["description"],
        price=p["price"],
        category=p["category"],
        brand=brand,
        features=[
            f"Brand: {brand}",
            f"Rating: {p.get('rating', 0)}/5",
            f"Stock: {p.get('stock', 0)} units",
        ],
        sold_by="Walmart",
        image_url=p.get("thumbnail")
        or (p.get("images") and p["images"][0])
        or "",
        inventory=[
            {
                "stock": p.get("stock", 0),
                "supplier": brand,
                "last_updated": datetime.utcnow(),
            }
        ],
        is_active=True,
        in_stock=p.get("stock", 0) > 0,
    )


def main():
    args = sys.argv[1:]
    clear = "--clear" in args
    count = int(parse_arg(args, "count", "50"))
    skip = int(parse_arg(args, "skip", "50"))

    if not os.getenv("MONGO_CONN"):
        print("MONGO_CONN not set in .env")
        sys.exit(1)

    connect(host=os.getenv("MONGO_CONN"))
    print("[seed] Connected to MongoDB")

    if clear:
        deleted = Product.objects.delete()
        print(f"[seed] Cleared {deleted} existing products")

    url = f"https://dummyjson.com/products?limit={count}&skip={skip}"
    print(f"[seed] Fetching {url}")
    resp = requests.get(url, timeout=15)
    resp.raise_for_status()
    products = resp.json().get("products", [])
    print(f"[seed] Got {len(products)} products")

    inserted = 0
    skipped = 0
    for p in products:
        try:
            doc = transform(p)
            if not doc.name or not doc.image_url:
                skipped += 1
                continue
            doc.save()
            inserted += 1
        except NotUniqueError:
            skipped += 1
        except Exception as e:
            print(f"[seed] Skipped '{p.get('title')}': {e}")
            skipped += 1

    print(f"[seed] Inserted {inserted} new, {skipped} duplicates/errors skipped")
    print("[seed] Done")


if __name__ == "__main__":
    main()

File: Amazon/seed.js
/**
 * Seeds the Amazon products collection from the public DummyJSON API.
 *
 * Usage:
 *   node seed.js                # add new products, skip duplicates
 *   node seed.js --clear        # delete all products first, then seed
 *   node seed.js --count=100    # how many products to fetch (max 100)
 *   node seed.js --skip=50      # offset (use to pull different items than Walmart)
 *
 * Requires Node 18+ (uses native fetch).
 * Reads MONGO_CONN from .env.
 */

require('dotenv').config();
const mongoose = require('mongoose');
const ProductsModel = require('./Models/Product');


const args = process.argv.slice(2);
const clear = args.includes('--clear');
const count = parseInt((args.find((a) => a.startsWith('--count=')) || '--count=50').split('=')[1], 10);
const skip = parseInt((args.find((a) => a.startsWith('--skip=')) || '--skip=0').split('=')[1], 10);


const transform = (p) => ({
    name: p.title,
    description: p.description,
    price: p.price,
    category: p.category,
    brand: p.brand || 'Generic',
    features: [
        p.brand ? `Brand: ${p.brand}` : `Category: ${p.category}`,
        `Rating: ${p.rating}/5`,
        `Stock: ${p.stock} units`
    ],
    soldBy: 'Amazon',
    imageUrl: p.thumbnail || (Array.isArray(p.images) && p.images[0]) || '',
    inventory: {
        stock: p.stock,
        supplier: p.brand || 'Default Supplier',
        lastUpdated: new Date()
    },
    isActive: true,
    inStock: p.stock > 0
});


async function seed() {
    if (!process.env.MONGO_CONN) {
        console.error('MONGO_CONN not set in .env');
        process.exit(1);
    }

    await mongoose.connect(process.env.MONGO_CONN);
    console.log('[seed] Connected to MongoDB');

    if (clear) {
        const { deletedCount } = await ProductsModel.deleteMany({});
        console.log(`[seed] Cleared ${deletedCount} existing products`);
    }

    const url = `https://dummyjson.com/products?limit=${count}&skip=${skip}`;
    console.log(`[seed] Fetching ${url}`);
    const resp = await fetch(url);
    if (!resp.ok) throw new Error(`DummyJSON returned ${resp.status}`);
    const { products } = await resp.json();
    console.log(`[seed] Got ${products.length} products`);

    const docs = products.map(transform).filter((d) => d.name && d.imageUrl);

    try {
        const result = await ProductsModel.insertMany(docs, { ordered: false });
        console.log(`[seed] Inserted ${result.length} products`);
    } catch (e) {
        const dupCount = (e.writeErrors || []).length;
        const inserted = (e.result && e.result.nInserted) || (e.insertedDocs && e.insertedDocs.length) || 0;
        console.log(`[seed] Inserted ${inserted} new, ${dupCount} duplicates skipped`);
    }

    await mongoose.disconnect();
    console.log('[seed] Done');
}


seed().catch((e) => {
    console.error('[seed] Failed:', e.message);
    process.exit(1);
});
